x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java

```diff
@@ -404,4 +444,76 @@ void doExecute(Action<Request, Response, RequestBuilder> action, Request request
+            this.followClient = followClient;
+            this.leaderIndex = leaderIndex;
+            this.followIndex = followIndex;
+            this.leaderClient = leaderClient;
+            this.currentIndexMetadataVersion = new AtomicLong();
+        }
+
+        public void accept(Long minimumRequiredIndexMetadataVersion, Consumer<Exception> handler) {
+            if (currentIndexMetadataVersion.get() >= minimumRequiredIndexMetadataVersion) {
+                LOGGER.debug("Current index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                    currentIndexMetadataVersion.get(), minimumRequiredIndexMetadataVersion);
+                handler.accept(null);
+            } else {
+                updateMapping(minimumRequiredIndexMetadataVersion, handler);
+            }
+        }
+
+        void updateMapping(long minimumRequiredIndexMetadataVersion, Consumer<Exception> handler) {
+            try {
+                updateMappingSemaphore.acquire();
+            } catch (InterruptedException e) {
+                handler.accept(e);
+                return;
+            }
+            if (currentIndexMetadataVersion.get() >= minimumRequiredIndexMetadataVersion) {
+                updateMappingSemaphore.release();
+                LOGGER.debug("Current index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                    currentIndexMetadataVersion.get(), minimumRequiredIndexMetadataVersion);
+                handler.accept(null);
+                return;
+            }
+
+            ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+            clusterStateRequest.clear();
+            clusterStateRequest.metaData(true);
+            clusterStateRequest.indices(leaderIndex.getName());
+
+            leaderClient.admin().cluster().state(clusterStateRequest, ActionListener.wrap(clusterStateResponse -> {
+                IndexMetaData indexMetaData = clusterStateResponse.getState().metaData().getIndexSafe(leaderIndex);
+                assert indexMetaData.getMappings().size() == 1;
+                MappingMetaData mappingMetaData = indexMetaData.getMappings().iterator().next().value;
+
+                PutMappingRequest putMappingRequest = new PutMappingRequest(followIndex.getName());
+                putMappingRequest.type(mappingMetaData.type());
+                putMappingRequest.source(mappingMetaData.source().string(), XContentType.JSON);
+                followClient.admin().indices().putMapping(putMappingRequest, ActionListener.wrap(putMappingResponse -> {
+                    currentIndexMetadataVersion.set(indexMetaData.getVersion());
+                    updateMappingSemaphore.release();
+                    handler.accept(null);
+                }, e -> {
+                    updateMappingSemaphore.release();
+                    handler.accept(e);
```

Comment:
a mapping collision here is not an error we should just retry.

Reply:
I don't think we should retry in case of a mapping collision. It means something bad has happened (most likely a bug in our side). Retrying wouldn't help, because the same error would occur in subsequent requests.

In case that several tasks update the mapping, no error should occur, only the first mapping update request would alter the mapping the other requests would be noops.